Jewish Calendar - support omer count after sunset

yohaybn opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Set language and date optional ('he' and today are defaults ), 
And make the service return the correct number after sunset


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/38706
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

amitfin commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -37,9 +41,27 @@
     async def get_omer_count(call: ServiceCall) -> ServiceResponse:
         """Return the Omer blessing for a given date."""
-        hebrew_date = HebrewDate.from_gdate(call.data["date"])
+        hebrew_date = get_hebrew_date(call.data["date"], call.hass)
```
Should be changed as it assumes "date" exists although it's now optional.

yohaybn replied:
if date is not supplied by user the default is today
from scheme  `vol.Optional(ATTR_DATE,` default=datetime.date.today): cv.date,`

amitfin commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -37,9 +41,27 @@
 def async_setup_services(hass: HomeAssistant) -> None:
     """Set up the Jewish Calendar services."""
 
+    def get_hebrew_date(date: datetime.date, hass: HomeAssistant) -> HebrewDate:
```
Maybe the sunset logic should be done only when "date" is not supplied.

yohaybn replied:
this is what I did in the code.
```  
# If the date is today, probabaly send default date
if date != today:
   return HebrewDate.from_gdate(date)
```

amitfin commented on homeassistant/components/jewish_calendar/services.yaml:
```diff
@@ -18,7 +18,7 @@ count_omer:
             - "italian"
     language:
-      required: true
+      required: false
```
The corresponding code should be changed accordingly (currently it assumes the parameter is supplied).

yohaybn replied:
I think Nusch need to be required. also in the code its Required
vol.Required(ATTR_NUSACH, default="sfarad"): vol.In(
            [nusach.name.lower() for nusach in Nusach]
        ),

tsvi reviewed (changes requested):
See my comment earlier

yohaybn commented:
> I think the correct approach here would be to allow the user to tick a checkbox whether they want the value of after_sunset (+ 1 day) or to return the value of today before sunset. This will make the code much easier to maintain without complex logic to calculate the current sunset. It makes also more sense in terms of sending a request and getting a response without considering the current context.

I think most of the users use this service in some type of automation that run in tzet hacochavim or in fixed time after sunset as reminder.
maybe add the checkbox as optional and default as true?
so user can send just his nusch and get the correct text

yohaybn commented:
maybe remove default date from schema and if not supplied use the current time with the sunset offset.
something like that

```python
        vol.Optional(ATTR_DATE): cv.date,
....

    async def get_omer_count(call: ServiceCall) -> ServiceResponse:
        """Return the Omer blessing for a given date."""
        if "date" in call.data:
            hebrew_date = HebrewDate.from_gdate(call.data["date"])
        else:
            hebrew_date = get_hebrew_date(call.hass)
        nusach = Nusach[call.data["nusach"].upper()]

```

aviadlevy commented:
> I think the correct approach here would be to allow the user to tick a checkbox whether they want the value of after_sunset (+ 1 day) or to return the value of today before sunset. This will make the code much easier to maintain without complex logic to calculate the current sunset. It makes also more sense in terms of sending a request and getting a response without considering the current context.

I’m not sure this is the most intuitive approach. The integration is closely tied to the concept of sunset — the strongest example being that the date changes at sunset.

tsvi commented:
@aviadlevy I think you're right that we should be tied to the Hebrew date which changes at sunset. But since the user is providing a non-hebrew date, it's best to get the indication wether they mean after sunset. It removes a lot of complexity in the code.

amitfin commented:
A common use-case is to use "now" for this service. It will make the service simpler to use if there will be a way to specify "now" by not providing the "date" parameter.

yohaybn committed:
Add 'is_after_sunset' option to count_omer service

Add 'is_after_sunset' option to count_omer service Add and tests for empty date

yohaybn committed:
Add tests for empty date

yohaybn committed:
Update service.py

tsvi commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -17,29 +18,56 @@
     {
-        vol.Required(ATTR_DATE, default=datetime.date.today): cv.date,
+        vol.Optional(ATTR_DATE): cv.date,
+        vol.Optional(AFTER_SUNSET, default=False): cv.boolean,
```
I'm not sure I'd use False as the default, since the count is usually done after sunset... Usually people relate to a given day as to what will be counted after sunset.

tsvi commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -17,29 +18,56 @@
 
 
+# top level for mocking tests
+def get_hebrew_date(hass: HomeAssistant) -> HebrewDate:
```
Maybe a cleaner way is to return today's date.
That way, the below logic becomes: 

```python
date = call.data.get("date", get_date())
after_sunset = call.data.get(AFTER_SUNSET, is_after_sunset())
hebrew_date = HebrewDate.from_gdate(date + timedelta(days=int(after_sunset))
```

This has an additional advantage to allow asking for after_sunset of today.
So we can have:
- a specific date 
- a specific date after sunset
- today (nothing specified)
- today after sunset (only after sunset specified)

The only "issue" is if someone asks after_sunset and it's after midnight where they probably meant simply today.

yohaybn committed:
Delete homeassistant/components/jewish_calendar/translations directory

frenck commented:
I've marked this PR a draft, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

yohaybn committed:
Update test_service.py

yohaybn committed:
set after sunset as true for default

yohaybn committed:
set after sunset as true for default and fix tests

tsvi commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -37,9 +42,34 @@
     async def get_omer_count(call: ServiceCall) -> ServiceResponse:
         """Return the Omer blessing for a given date."""
-        hebrew_date = HebrewDate.from_gdate(call.data["date"])
+        if "date" in call.data:
+            date = call.data["date"]
+            hebrew_date = HebrewDate.from_gdate(
+                date + datetime.timedelta(days=int(call.data["is_after_sunset"]))
+            )
+        else:
+            hebrew_date = get_hebrew_date(call.hass)
```
```suggestion
        today = dt_util.now().date()
        date = call.data.get("date", today)
        after_sunset = call.data[AFTER_SUNSET] if date != today else is_after_sunset(hass, today)
        hebrew_date = HebrewDate.from_gdate(date + datetime.timedelta(days=int(after_sunset)))
```

tsvi commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -37,9 +42,34 @@
+            dt_util.now() + datetime.timedelta(days=int(after_sunset))
+        )
+
+    def is_after_sunset(hass: HomeAssistant) -> bool:
+        """Determine if the current time is after sunset."""
+        now = dt_util.now()
+        today = now.date()
+        event_date = get_astral_event_date(hass, SUN_EVENT_SUNSET, today)
```
```suggestion
    def is_after_sunset(hass: HomeAssistant, date: datetime.date) -> bool:
        """Determine if the current time is after sunset for a given date."""
        event_date = get_astral_event_date(hass, SUN_EVENT_SUNSET, date)
```

tsvi commented on homeassistant/components/jewish_calendar/service.py:
```diff
@@ -37,9 +42,34 @@
 def async_setup_services(hass: HomeAssistant) -> None:
     """Set up the Jewish Calendar services."""
 
+    def get_hebrew_date(hass: HomeAssistant) -> HebrewDate:
+        """Convert a Gregorian date to a Hebrew date."""
+        after_sunset = is_after_sunset(hass)
+        return HebrewDate.from_gdate(
+            dt_util.now() + datetime.timedelta(days=int(after_sunset))
+        )
```
Can be removed, see below

tsvi commented on tests/components/jewish_calendar/test_service.py:
```diff
@@ -12,44 +12,94 @@
-    test_date: dt.date,
-    nusach: str,
-    language: Language,
+    test_time: dt.date | None,
```
test_time is dt.datetime

tsvi commented on tests/components/jewish_calendar/test_service.py:
```diff
@@ -12,44 +12,94 @@
+            return_response=True,
+        )
+
+    if test_time:
+        with freeze_time(test_time):
+            result = await call_service()
+    else:
```
Can be removed

tsvi commented on tests/components/jewish_calendar/test_service.py:
```diff
@@ -12,44 +12,94 @@
     expected: str,
 ) -> None:
     """Test get omer blessing."""
     config_entry.add_to_hass(hass)
     await hass.config_entries.async_setup(config_entry.entry_id)
     await hass.async_block_till_done()
```
Can be removed

tsvi commented on tests/components/jewish_calendar/test_service.py:
```diff
@@ -12,44 +12,94 @@
         ),
     ],
+    indirect=["test_time"],
 )
```
Add:
```suggestion
)
@pytest.mark.usefixture("setup_at_time")
```

to remove the initial setup, as well as the need for using freezegun

tsvi committed:
Fix the tests using setup_at_time fixture

tsvi committed:
Rename is_after_sunset to after_sunset and use constant

tsvi committed:
Simplify logic for hebrew_date calculation

tsvi committed:
Merge branch 'dev' of https://github.com/home-assistant/core into pr/yohaybn/143332
